src/com/google/enterprise/adaptor/database/DatabaseAdaptor.java

```diff
@@ -693,22 +689,36 @@ public static void main(String[] args) {
     AbstractAdaptor.main(new DatabaseAdaptor(), args);
   }
   
-  private Acl getAcl(Connection conn, String uniqueId) throws SQLException {
-    try (PreparedStatement stmt = getAclFromDb(conn, uniqueId);
-        ResultSet rs = stmt.executeQuery()) {
-      log.finer("got acl");
-      return buildAcl(rs, aclPrincipalDelimiter, aclNamespace);
+  private Acl getAcl(Connection conn, ResultSet rs, String uniqueId)
+      throws SQLException {
+    if (aclSql != null || rs == null) {
+      try (PreparedStatement stmt = getAclFromDb(conn, uniqueId);
+          ResultSet resSet = stmt.executeQuery()) {
+        log.finer("got acl");
+        boolean hasResult = resSet.next();
+        if (!hasResult) {
+          // empty Acl ensures adaptor will mark this document as secure
+          return Acl.EMPTY;
+        } else {
+          if (aclSql != null) {
+            return buildAcl(resSet, aclPrincipalDelimiter, aclNamespace,
+                ResultSetNext.PROCESS_ALL_ROWS, Acl.EMPTY);
+          } else {
+            return buildAcl(resSet, aclPrincipalDelimiter, aclNamespace,
+                ResultSetNext.PROCESS_ALL_ROWS, null);
```

Comment:
This should be PROCESS_ONE_ROW. We're back to 3 checks for aclSql == null, which I think suggests we need some refactoring. Maybe some kind of ACL strategy pattern, where we construct an appropriate object once for each scenario?

Reply:
Yes, thx. Done.
Yes, agree, refactoring may reduce the null check for aclSql and/or look cleaner.